src/app/auth.service.ts

```diff
@@ -45,17 +51,21 @@ export class AuthService {
       );
   }
 
-  /** Authenticates a user and returns an Observable for their ID token. */
+  /** Authenticates a user, if not signed in, and returns an Observable for their ID token. */
   authenticateWithGoogle(): Observable<string> {
-    const authPromise = this.auth.signInWithPopup(
-      new auth.GoogleAuthProvider()
-    );
-    return from(authPromise).pipe(
-      flatMap(() => {
-        return this.auth.currentUser.then(user => {
-          if (user == null) throw new Error('No user found after signed in.');
-          return user.getIdToken(true);
-        });
+    return from(this.user).pipe(
+      flatMap(user => {
+        const authPromise = user
+          ? Promise.resolve()
+          : this.auth.signInWithPopup(new auth.GoogleAuthProvider());
+        return from(authPromise).pipe(
```

Comment:
This might be more readable as a guard clause:

```suggestion
        if (!user) return Promise.resolve();
        const authPromise = this.auth.signInWithPopup(new auth.GoogleAuthProvider());
        return from(authPromise).pipe(
```

Reply:
I still need to reach the final return, since I still need to return the user's ID token after determining if they need to be signed-in or not.

```suggestion
        const authPromise =
                !user ? this.auth.signInWithPopup(new auth.GoogleAuthProvider())
                          : Promise.resolve();
        return from(authPromise).pipe(
```